slang.h

```diff
@@ -2543,8 +2597,329 @@ namespace slang
+
+            /** Set the default matrix layout mode to use for all compiles.
+            */
+        virtual SLANG_NO_THROW SlangResult SLANG_MCALL setMatrixLayoutMode(
+            SlangMatrixLayoutMode mode) = 0;
+#endif
+    };
+
+    #define SLANG_UUID_ISession { 0x67618701, 0xd116, 0x468f, { 0xab, 0x3b, 0x47, 0x4b, 0xed, 0xce, 0xe, 0x3d } }
+
+
+        /** A module is the granularity of shader code compilation and loading.
+
+        In most cases a module corresponds to a single compile "translation unit."
+        This will often be a single `.slang` or `.hlsl` file and everything it
+        `#include`s.
+
+        Notably, a module `M` does *not* include the things it `import`s, as these
+        as distinct modules that `M` depends on. There is a directed graph of
+        module dependencies, and all modules in the graph must belong to the
+        same session (`ISession`).
+        */
+    struct IModule : public ISlangUnknown
+    {
+    public:
+    };
+    
+    #define SLANG_UUID_IModule { 0xc720e64, 0x8722, 0x4d31, { 0x89, 0x90, 0x63, 0x8a, 0x98, 0xb1, 0xc2, 0x79 } }
+
+
+        /** Argument used for specialization to types/values.
+        */
+    struct SpecializationArg
+    {
+        enum class Kind : int32_t
+        {
+            Unknown,
+            Type,
+        };
+        Kind kind;
+        union
+        {
+            TypeReflection* type;
+        };
+    };
+
+        /** Description of a program to be created.
+        */
+    struct ProgramDesc
+    {
+        struct Item
+        {
+            enum class Kind : int32_t
+            {
+                Program,
+                Module,
+            };
+            Kind kind;
+            union
+            {
+                IProgram*  program;
+                IModule*   module;
+            };
+        };
+
+        Item const* items;
+        SlangInt    itemCount;
+    };
+
+
+        /** A program comprises zero or more modules, entry points, etc. that have been linked together.
+        */
+    struct IProgram : public ISlangUnknown
+    {
+    public:
+            /** Get the runtime session that this program belongs to.
+            */
+        virtual SLANG_NO_THROW ISession* SLANG_MCALL getSession() = 0;
+
+            /** Get the layout for this program for the chosen `targetIndex`
+            */
+        virtual SLANG_NO_THROW ProgramLayout* SLANG_MCALL getLayout(
+            SlangInt    targetIndex = 0,
+            IBlob**     outDiagnostics = nullptr) = 0;
+
+            /** Get the compiled code for the entry point at `entryPointIndex` for the chosen `targetIndex`
+
+            If code has not already been generated for the given entry point and target,
+            then a compilation error may be detected, in which case `outDiagnostics`
+            (if non-null) will be filled in with a blob of messages diagnosing the error.
+            */
+        virtual SLANG_NO_THROW SlangResult SLANG_MCALL getEntryPointCode(
```

Comment:
So this is lazily compiling? If so it might be better to have verb other than 'get' to make clear it could do quite a bit of work.

Reply:
Yes, the idea is that you can in principle use the Slang API as your "shader cache" layer, and if/when we support loading from binary container files that could include pre-computed kernels, a naive app that does aggressive on-demand compilation could Just Work with our caching layer.

I agree that the naming is maybe under-selling the amount of work. At the implementation level this is a `getOrCreate...` instead of just `get...` which is more accurate, but clumsy.

I don't want the verb to imply that code generation happens every time.